### ISSUE TITLE
☂️ Website Redesign and Documentation Rewrite

### ISSUE THREAD

**Issue by cpojer** [facebook/react-native-website#929]
The content and look of the React Native website are outdated. Many of the existing guides and the overall flow of the documentation were written in 2015 and do not reflect well what it is like to build apps with React Native in 2019. When we asked what people [dislike about React Native](https://github.com/react-native-community/discussions-and-proposals/issues/104), documentation issues came up many times. We received [more concrete feedback](https://github.com/facebook/react-native-website/issues/917) when we asked specifically.

This umbrella issue aims to outline the work necessary to update the website both with a new look and new content. While we'll handle the redesign work outside of this issue, we'd love to get your help in updating the entire documentation: if there is any item below issue that you'd love to help with, please send a Pull Request to this repo or get in touch otherwise! The best way to help right now is by updating the API docs with the actual implementations (see below) to ensure things are up-to-date for the upcoming release.

# Goals

* A fresh look for React Native that represents the framework in 2019.
* Guides for beginners and advanced developers that are best in class.
* Always up-to-date API documentation that is synced with actual code and examples.

# Redesign

The redesign is currently being worked on by @orta and @tesseralis. We are excited for you all to get a look very soon!

# New Guide Structure
The current structure is quite outdated. It's hard to know upfront exactly what the new flow should be like but here is the proposal for a new structure which is open for discussion:

* **Basics**
  * Setup & Getting Started
  * Basics & React – this should mirror reactjs.org with a focus on RN, gently introduce RN and React, and link to reactjs.org for concepts.
  * ~~Props~~ – Removed and rolled into basics, with links to reactjs.org.
  * ~~State~~ – Removed and rolled into basics, with links to reactjs.org.
  * Style, Height and Width – Rolled up into a new guide "Styling & Dimensions"
  * Layout or "Thinking in flexbox" – Significantly expand this guide with more screenshots and possibly a live playground similar to yogalayout.com (first improvements were merged in https://github.com/facebook/react-native-website/pull/947).
  * Handling Text Input and Handling Touches – Merge into a single "Handling User Interactions" guide
  * Using a ScrollView and Using a List Views – Merge into a single guide
  * ~~Networking~~ – Ideally we'll factor data fetching into the preceding guides, like populating a ScrollView content, for example.
* **New Guides**
  * Guide on "The basics of native development for Android and iOS" aimed at web developers including Xcode, Android Studio and Cocoapods and Gradle
  * Include docs on RN integration into Kotlin and Swift apps.
  * A list of links on app deployment for Android and iOS
  * A guide on how to build a quality app (including links with information to app icons, launch screen, user interface guidelines and other polish).
* Improve the guide on building native modules including how to publish it as a package.
* Recommend third-party navigation libraries in the Navigation guide
* **All other existing guides**
  * Rethink all the guides beyond the basics, and improve them. This umbrella issue will be updated accordingly as we complete all the other work outlined in this issue.
  * RN Accessibility Guide
* ???

# Documentation Rewrite Tasks

* [x] ~~Remove the Contributing guide from the website. Move all RN development related content into the RN GitHub repo. (@hramos)~~
* [x] ~~Remove obsolete API docs (@cpojer)~~
* [x] ~~Make the high level sections in the sidebar collapsible. https://github.com/facebook/react-native-website/pull/934 (@yangshun)~~
* [x] ~~Get rid of the “Guidelines for Text (https://github.com/facebook/react-native-website/blob/master/CONTRIBUTING.md#guidelines-for-text)” (@cpojer)~~
* [x] ~~Minor: Remove TimerMixins from the Timer docs, it's already gone from the code.~~
* [x] ~~Minor: Get rid of packager/config.js stuff across the docs and recommend using the Metro file in the template.~~
* [x] ~~Split the website up into high-level entry points for Docs (Guides) and APIs (Components and API docs). https://github.com/facebook/react-native-website/pull/962 @yangshun~~
* [ ] Minor: AppRegistry lifecycle docs aren't good: https://github.com/react-native-community/discussions-and-proposals/issues/64#issuecomment-444868735
* [ ] A page that lists 30 of the top React Native community packages and high-quality third-party packages, like UI libraries that help people solve problems when building apps. See https://github.com/facebook/react-native-website/pull/884

# API Docs
This is one where you can help even if you only have a few minutes. We need to update all our API docs and align them with the actual APIs in the code. For example, if you'd like to ensure consistency for the `ScrollView` component, it's best to go to the react-native repo and look for the [ScrollView.js](https://github.com/facebook/react-native/blob/master/Libraries/Components/ScrollView/ScrollView.js) file. Then go through the ScrollView.md documentation in this repository and compare each prop that it accepts and update the `.md` file in this repo to align with what is in the code and send a Pull Request!

* [ ] **Components**
  * [x] ActivityIndicator
  * [x] Button
  * [x] DatePickerIOS
  * [ ] DrawerLayoutAndroid
  * [x] FlatList
  * [ ] Image
  * [x] ImageBackground
  * [x] InputAccessoryView
  * [x] KeyboardAvoidingView
  * [ ] Modal
  * [x] Picker
  * [ ] PickerIOS
  * [x] ProgressBarAndroid
  * [x] ProgressViewIOS
  * [x] RefreshControl
  * [x] SafeAreaView
  * [x] ScrollView
  * [ ] SectionList
  * [x] SegmentedControlIOS
  * [x] Slider
  * [x] StatusBar
  * [x] Switch
  * [x] Text
  * [ ] TextInput
  * [ ] ToolbarAndroid
  * [ ] TouchableHighlight
  * [ ] TouchableNativeFeedback
  * [x] TouchableOpacity
  * [x] TouchableWithoutFeedback
  * [x] View
  * [ ] ViewPagerAndroid
  * [ ] VirtualizedList
* [ ] **APIs**
  * [x] AccessibilityInfo
  * [x] ActionSheetIOS
  * [x] Alert
  * [ ] AlertIOS
  * [ ] Animated
  * [ ] AppRegistry
  * [ ] AppState
  * [x] AsyncStorage
  * [x] BackHandler
  * [ ] CameraRoll
  * [x] Clipboard
  * [x] DatePickerAndroid
  * [x] Dimensions
  * [x] Easing
  * [ ] ImageEditor
  * [x] ImagePickerIOS
  * [ ] Image Style Props
  * [x] InteractionManager
  * [ ] Keyboard
  * [ ] Layout Props
  * [x] LayoutAnimation
  * [ ] Linking
  * [ ] ListViewDataSource
  * [ ] PanResponder
  * [x] PermissionsAndroid
  * [x] PixelRatio
  * [ ] PushNotificationIOS
  * [x] Settings
  * [x] Shadow Props
  * [x] Share
  * [x] StatusBarIOS
  * [ ] StyleSheet
  * [x] Systrace
  * [x] Text Style Props
  * [x] TimePickerAndroid
  * [x] ToastAndroid
  * [x] Transforms
  * [x] Vibration
  * [ ] View Style Props

# Guidelines
* We'll prioritize improving high-impact content based on users per page per month. Right now, the pages with the most views are the guides in the basics section so we'll prioritize those.
* The getting started tutorial should be a full guide from first principles to getting started, building, linking and deploying an application. It should obviate the need for a getting started guide anywhere else.
* The documentation from a high-level should include all necessary details to build a performant high-quality application with React Native.
* Add live demos, screenshots or find other ways to engage with the content. This could also come in the form of a real example app, although we have not had the time to build one unfortunately. If the introduction was about building a real app and was a connected step-by-step guide, that would be ideal.
* If there is documentation that already exists in an official place (like Android documentation) elsewhere, link to it instead of recreating documentation. The website should be hyper focused on React Native itself. An example of this is linking to other guides on how to deploy an app to the app store, unless it requires specific knowledge because of React Native.

Finally, we are thinking of merging this repo back into the React Native repo to ensure API documentation will always stay up-to-date going forward. For now we'll keep using this repository for the purpose of this umbrella issue and we will attempt the merge later.

**Comment by glauberfc** [comment]
Can this website be made w/ Gatsby?

**Comment by cpojer** [comment]
We are not considering a technical rewrite of the code. Docusauraus works fine for us at this time.

**Comment by saxenanickk** [comment]
Would love to contribute in this.

**Comment by glauberfc** [comment]
@cpojer I'm comparing `ScrollView.js` and `ScrollView.md` files and I have some questions:

1. Should all props inside `type IOSProps` be annotated w/ `@platform ios`. The prop below hasn't this annotation on the file:

>   /**
>    * When true, shows a horizontal scroll indicator.
>    * The default value is true.
>    */
>   showsHorizontalScrollIndicator?: ?boolean,

2. Should the props be listed on alphabetical order?

3. In this part of the docs there is a link for View props indicating that this component enhance these props, but there is not a link for TouchableProps. 
<img width="717" alt="Screen Shot 2019-05-15 at 12 13 59 PM" src="https://user-images.githubusercontent.com/15311858/57787011-fbd61380-770a-11e9-81ca-20ad468a4374.png">
Do we need to create some link that show the TouchableProps too?

**Comment by cpojer** [comment]
@glauberfc:
* Yes, IOSProps are usually ios specific.
* Yes, alphabetical order is ideal.
* You are right that seems to be missing. Wanna send a PR just for the touchable props?

**Comment by glauberfc** [comment]
> You are right that seems to be missing. Wanna send a PR just for the touchable props?

@cpojer Do I need only create a file named `TouchableProps.md`? Is there some text explaining each of these props?

```
type TouchableProps = $ReadOnly<{|
  onTouchStart?: (event: PressEvent) => void,
  onTouchMove?: (event: PressEvent) => void,
  onTouchEnd?: (event: PressEvent) => void,
  onTouchCancel?: (event: PressEvent) => void,
  onTouchEndCapture?: (event: PressEvent) => void,
|}>;
```

**Comment by ndejesus1227** [comment]
I got dibs on View and Text!

**Comment by yangshun** [comment]
@glauberfc would love to know your reasons for moving to Gatsby 😄

**Comment by glauberfc** [comment]
@yangshun I thought the new React site was made with Gatsby. I think this because all sites made w/ Docusaurus are very similarly and the React website is a bit more "styled"

**Comment by ndejesus1227** [comment]
So I made a PR for View, do my eyes deceive me or is there literally no `Text.js`???

**Comment by orta** [comment]
>  Make the high level sections in the sidebar collapsible.

I think this is 'just' updating docusaurus BTW

**Comment by yangshun** [comment]
> I think this is 'just' updating docusaurus BTW

Yep there's now a siteConfig option called `docsSideNavCollapsible` and you can set it to `true`.

**Comment by ndejesus1227** [comment]
@yangshun @orta Sorta related to gatsby, I think it would be cool to implement MDX for handling these docs. 


Every section for an API could be a component instead, should be much easier to reason about/read/contribute that way :O

**Comment by yangshun** [comment]
@ndejesus1227 Docusaurus 2 will support MDX. Unfortunately we haven't implementing versioning and hence cannot port React Native website to Docusaurus 2. Check out https://endiliey.com/2019/05/16/docusaurus-2-technical-preview/ for a preview.

**Comment by AbdelkhalekESI** [comment]
Hi guys , i'd like to contribute , any contribution guide available and from where we should start?

**Comment by cpojer** [comment]
@yangshun any chance you could help us update Docusaurus so we can make the sidebar collapsible?

**Comment by ksrini73** [comment]
I would love to contribute on this.

**Comment by glauberfc** [comment]
@cpojer On [ScrollView](https://github.com/facebook/react-native/blob/master/Libraries/Components/ScrollView/ScrollView.js) the prop `showsHorizontalScrollIndicator` hasn't the `@platform ios` annotation, but are inside `type IOSProps`. Is it a missing or is it right?

**Comment by cpojer** [comment]
@glauberfc 
1. Seems like it's available on Android: https://github.com/facebook/react-native/blob/faaa92bb04467f1d8cef0fe7ce309643486b7a0d/ReactAndroid/src/main/java/com/facebook/react/views/scroll/ReactHorizontalScrollViewManager.java#L73
2. I would assume we have both, one is the boolean to toggle the behavior the other is an event that is run when scrolling to top.

**Comment by glauberfc** [comment]
> Sorta related to gatsby, I think it would be cool to implement MDX for handling these docs.

I agree w/ you @ndejesus1227. I would like there is a way of import the component file and document it w/ MDX like in [Docz](https://www.docz.site/). So, this will avoid us to copy and paste many times and will keep the docs updated.

**Comment by SSTPIERRE2** [comment]
Having worked with RN in it's early stages and learning from the older documentation I find this super exciting! The changes outlines here look excellent 🤘

**Comment by yangshun** [comment]
@cpojer will do so! I'll also enable a bunch of other useful defaults and see if you guys like it.

**Comment by glauberfc** [comment]
If there was something like [this](https://www.youtube.com/watch?v=Yl6WARA3IhQ) on Docusaurus would help a lot on the API updates. People would use a pattern to document all APIs and the JS file would be converted on a markdown file

**Comment by yangshun** [comment]
@glauberfc That would be possible with plugins in Docusaurus 2. I have to add a backward-compatible versioning ability to D2 before we can move RN website to D2.

**Comment by ndejesus1227** [comment]
I took care of View, I can't seem to find Text.js so I'm gonna pass on that one.

I'd like to call dibs on Button and FlatList! (Going for Button first)

**Comment by raajnadar** [comment]
@cpojer I went through the Vibration docs and found no changes and didn't find any file for `Transform.js`

**Comment by ecreeth** [comment]
#947

**Comment by MohamadKh75** [comment]
I want to create a PR and sort all of the guides (iOS and Android too) alphabetically. I won't change "The Basics" category as it needs to be sorted in the other way!
Shall i start or it can be done by Docusaurus?

**Comment by hramos** [comment]
I'd prefer the order of guides to be intentional. Think of it as a story. What should the progression look like as you ramp up on React Native?

**Comment by raajnadar** [comment]
cc @cpojer can you update the original post and also check my last comment https://github.com/facebook/react-native-website/issues/929#issuecomment-494348891

**Comment by ecreeth** [comment]
@cpojer The properties of this component has not changed: [ActivityIndicator.md](https://github.com/facebook/react-native/blob/master/Libraries/Components/ActivityIndicator/ActivityIndicator.js)

**Comment by ecreeth** [comment]
@cpojer The properties of this component has not changed: [RefreshControl.md](https://github.com/facebook/react-native/blob/master/Libraries/Components/RefreshControl/RefreshControl.js)

**Comment by ecreeth** [comment]
@cpojer The properties of this component has not changed: [Picker.md](https://github.com/facebook/react-native/blob/master/Libraries/Components/Picker/Picker.js)

**Comment by ecreeth** [comment]
@cpojer The properties of this component has not changed: [PickerItem.md](https://github.com/facebook/react-native/blob/master/Libraries/Components/Picker/Picker.js#L25)

**Comment by ecreeth** [comment]
@cpojer The properties of this component has not changed: [ProgressViewIOS](https://github.com/facebook/react-native/blob/master/Libraries/Components/ProgressViewIOS/ProgressViewIOS.ios.js)

**Comment by ecreeth** [comment]
@cpojer The properties of this component has not changed: [SafeAreaView](https://github.com/facebook/react-native/blob/master/Libraries/Components/SafeAreaView/SafeAreaView.js)

**Comment by ecreeth** [comment]
@cpojer The properties of this component has not changed: [SegmentedControlIOS](https://github.com/facebook/react-native/blob/master/Libraries/Components/SegmentedControlIOS/RCTSegmentedControlNativeComponent.js)

**Comment by yangshun** [comment]
@ecreeth Consider consolidating your updates in one single post/comment and updating it when there are new things to add!

**Comment by ecreeth** [comment]
@cpojer The properties of this component has not changed: 
- [SegmentedControlIOS](https://github.com/facebook/react-native/blob/master/Libraries/Components/SegmentedControlIOS/RCTSegmentedControlNativeComponent.js)
- [SafeAreaView](https://github.com/facebook/react-native/blob/master/Libraries/Components/SafeAreaView/SafeAreaView.js)
- [ProgressViewIOS](https://github.com/facebook/react-native/blob/master/Libraries/Components/ProgressViewIOS/ProgressViewIOS.ios.js)

**Comment by cpojer** [comment]
@ecreeth thank you so much for doing this! I just updated the list above :)

**Comment by ecreeth** [comment]
@cpojer The following **components** or **API's** have been removed or merged:

- `AlertIOS` see [#23318](https://github.com/facebook/react-native/pull/23318)

**Comment by cpojer** [comment]
@ecreeth can you remove these docs via a PR?

**Comment by ecreeth** [comment]
> @ecreeth can you remove these docs via a PR?

Of course yes!

**Comment by SaraChicaD** [comment]
On the `Slider` [repo](https://github.com/react-native-community/react-native-slider) the [Slider.js](https://github.com/react-native-community/react-native-slider/blob/master/js/Slider.js) file matches the [docs](https://github.com/react-native-community/react-native-slider#props). 

On the RN web docs, which I understand are deprecated, `onSlidingStart` is [missing](https://facebook.github.io/react-native/docs/slider.html).

**Comment by jeremyscatigna** [comment]
Would love to help! 🙌What could I do ?

**Comment by mikeaustin** [comment]
What I don't understand is how the core method of registering and running an app is not documented? AppRegistry.runApplication() has 0 documentation in https://facebook.github.io/react-native/docs/appregistry#runapplication. I Googled to find out what appParameters are supported, but actually didn't find much. Is there not a transition period where the docs are not empty?

**Comment by orta** [comment]
This sounds like a great opportunity for you to try and contribute to the docs @mikeaustin (/cc @jeremyscatigna  also?) - looks like these are [the params](https://github.com/facebook/react-native/blob/master/Libraries/ReactNative/AppRegistry.js#L36-L41) and here's [the function](https://github.com/facebook/react-native/blob/master/Libraries/ReactNative/AppRegistry.js#L81)

**Comment by ecreeth** [comment]
@cpojer Please update the list above with the PRs that were merged. 📦 💯 To know better what we are missing

**Comment by BatDroid** [comment]
I'm really interested in contributing to this repo. i've started working on `SectionList` but as a starter i have couple of questions that i would really appreciate if someone could help me with.
1. is it good if i re-order the props alphabetically?
2. `renderItem` is Optional in codebase. The reason is the functionality of rendering per item behaviour that if user use it then the `renderItem` will not be needed anymore. should i move it to optional props?
3. based on what is in the codebase about the `legacyImplementation`, `The legacy implementation is no longer supported.`. shall i add it to docs? (or there has been a reason that it isn't)
4. based on the codebase, the `flashScrollIndicators` seems to be only for `ios`. shall i add it to docs? (or there has been a reason that it hasn't been mentioned)
really sorry for these questions but if i get more familiar with these things, i'll do things faster. 
ill make my pr after i get these answers. thank you

**Comment by keshavkaul** [comment]
@cpojer Can someone review this PR that changes the recommended way to setup a new react native library? The currently recommended way is to use `react-native-create-library` package. But this package is not being maintained. The PR recommends using `create-react-native-module` instead.

https://github.com/facebook/react-native-website/pull/1118

**Comment by rachelnabors** [comment]
We've moved the remaining tasks into a new ticket: https://github.com/facebook/react-native-website/issues/1579 Please go there to help us finish up the component docs and thanks for all your help thus far!
